# Prostory jmen a obor platnosti proměnných:
- Upravte funkci nalézající prvočísla tak, aby ukládala již spočítané prvočísla do listu v defaultním argumentu funkce. 


In [5]:
def primes(n: int, _primes = [2]) -> list:
    if len(_primes) > n:
        return _primes[:n]
    cur = _primes[-1]
    while len(_primes) < n:
        cur += 1
        for i in _primes:
            if not cur % i:
                break
        else:
            _primes.append(cur)

    return _primes

print(primes(10))
print(primes(30))
print(primes(15))

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29]
[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113]
[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47]



# Jupyter magics:
- použijte funkci %%writefile a pomocí ní (a kopie funkcí z posledního DU) vytvořte .py soubor sloužící jako modul mající tři funkce viz minulý DU
- Použijte magic příkazy %time %timeit %%time a %%timeit k měření času běhu různých částí kódu (například funkcí z minulého DU).


In [6]:
%%writefile module_02.py
def iter_sqrt(a: float, n: int) -> float:
    """Calculates the approximation of the square root of a given number using the simple iteration method.

    Input parameters:
    a - value to get square root of
    n - number of iterations

    Outputs:
    approximation of square root of a
    """
    #checks
    if a < 0:
        print("Err: square root doesn't accept negative values")
        return 0
    elif a == 0:
        return 0
    if n <= 0:
        print("Err: the number of iterations should be larger than 0")
        return a
    
    #method
    x = a
    for i in range(0, n):
        x = (a/x + x)/2
    return x

def iter_pi(n: int) -> float:
    """Approximates pi by using the regular n-gon method.

    Input parameter:
    n - number of iterations

    Outputs:
    approximation of pi
    """
    #checks
    if n <= 0:
        print("Err: the number of iterations should be larger than 0")
        return 0
    
    #method
    l = 1
    h = iter_sqrt(.75, 100)
    for i in range(1, n):
        l = iter_sqrt((l/2)**2 + (1-h)**2, 100)
        h = iter_sqrt(1 - (l/2)**2, 100)

    return 1.5 * 2**n * l * h

def iter_newton(n: int) -> float:
    """Approximates pi by using Newton's method.

    Input parameter:
    n - number of iterations

    Outputs:
    approximation of pi
    """
    #checks
    if n <= 0:
        print("Err: the number of iterations should be larger than 0")
        return 0
    
    a = 1/16
    sum = a / 3
    for i in range(2, n+1):
        a *= (2*i - 3) / (i * 2**3)
        sum += a / (2*i + 1)
    return 12 * (.5 - iter_sqrt(3, 1000)/8 - sum)


Writing module_02.py


In [7]:
from module_02 import iter_sqrt, iter_pi, iter_newton

In [8]:
%time
iter_sqrt(2, 100)
%time
iter_pi(50)
%time
iter_pi(50)

CPU times: user 3 µs, sys: 0 ns, total: 3 µs
Wall time: 9.06 µs
CPU times: user 2 µs, sys: 0 ns, total: 2 µs
Wall time: 7.63 µs
CPU times: user 1 µs, sys: 0 ns, total: 1 µs
Wall time: 3.34 µs


3.1415926535897922

In [9]:
%%prun
iter_pi(50)

         103 function calls in 0.001 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
       99    0.001    0.000    0.001    0.000 module_02.py:1(iter_sqrt)
        1    0.000    0.000    0.001    0.001 module_02.py:27(iter_pi)
        1    0.000    0.000    0.001    0.001 {built-in method builtins.exec}
        1    0.000    0.000    0.001    0.001 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}

In [10]:
%%prun
iter_newton(50)

         5 function calls in 0.000 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.000    0.000 module_02.py:1(iter_sqrt)
        1    0.000    0.000    0.000    0.000 module_02.py:50(iter_newton)
        1    0.000    0.000    0.000    0.000 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}


# Práce s řetězci:

- Použíjte funkce z minulého úkolu a v cyklu spočtěte $\pi$ s různým $n$. Pomocí f-strings vypište výsledky ve formátu "n = 5, pi = 3.14159, počet správných cifer = 5".


In [11]:
import math
precision = 15
math_pi = f"{math.pi:1.{precision}f}"
def pi_string(n, func):
    return f"{func(n):1.{precision}f}"
def get_digits(str):
    for i in range(precision):
        if str[i+2] != math_pi[i+2]:
            return i
    return precision

print(f"\tpi = {math_pi}")
for i in range(1, 21):
    pi = pi_string(i, iter_newton)
    print(f"n = {i:>2}, pi = {pi}, počet správných cifer = {get_digits(pi):>2}")

	pi = 3.141592653589793
n =  1, pi = 3.151923788646684, počet správných cifer =  1
n =  2, pi = 3.142548788646685, počet správných cifer =  2
n =  3, pi = 3.141711735075256, počet správných cifer =  3
n =  4, pi = 3.141610009814839, počet správných cifer =  3
n =  5, pi = 3.141595444607097, počet správných cifer =  5
n =  6, pi = 3.141593133780869, počet správných cifer =  5
n =  7, pi = 3.141592740390214, počet správných cifer =  6
n =  8, pi = 3.141592669883616, počet správných cifer =  7
n =  9, pi = 3.141592656740940, počet správných cifer =  8
n = 10, pi = 3.141592654214103, počet správných cifer =  8
n = 11, pi = 3.141592653715977, počet správných cifer =  9
n = 12, pi = 3.141592653615729, počet správných cifer =  9
n = 13, pi = 3.141592653595201, počet správných cifer = 10
n = 14, pi = 3.141592653590934, počet správných cifer = 10
n = 15, pi = 3.141592653590036, počet správných cifer = 10
n = 16, pi = 3.141592653589846, počet správných cifer = 12
n = 17, pi = 3.141592653589805, 


# Generátory a iterátory:

- Napište generátor (funkci), který vytváří Fibonacciho posloupnost a vrátí prvních $n$ členů.


In [12]:
def fibonacci():
    a = 0
    b = 1
    yield 1

    while True:
        c = a + b
        yield c
        a = b
        b = c

gen = fibonacci()

n = 10
print([next(gen) for i in range(0, n)])

[1, 1, 2, 3, 5, 8, 13, 21, 34, 55]



# List comprehension:
- Vytvořte seznam čísel od 1 do 10 pomocí list comprehension.
- Vytvořte seznam druhých mocnin čísel od 1 do 10 pomocí list comprehension.
- Vytvořte seznam prvočísel od 1 do 100 pomocí list comprehension. Hint (použijte funkci all() na generátor testující dělitelnost).


In [13]:
list1 = list(range(1, 11))
list2 = list(i**2 for i in range(1, 11))
list3 = list(i for i in range(2, 101) if all(i % p for p in range(2, int(i**0.5) + 1)))

print(f"{list1}\n{list2}\n{list3}")

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]
[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97]



# Práce se soubory:
- Pomocí %%writefile vytvořte soubor s nějakým textem.
- Otevřete soubor pro čtení a vypište jeho obsah.
- Otevřete jiný soubor pro zápis a napište do něj textový řetězec.
- Pomocí f-strings zapište do souboru tabulku čísel od 1 do 10, jejich druhých mocnin a jejich druhých odmocnin.

In [14]:
%%writefile text_02.txt
text text tady je text
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa

Writing text_02.txt


In [15]:
try:
    with open("text_02.txt", "r") as f:
        for line in f:
            print(line)
except FileNotFoundError:
    print("Error reading file")

text text tady je text

aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa



In [16]:
with open("table_02", "w") as f:
    f.write(f"{'n':>7},{'n*n':>7},{'sqrt(n)':>8}\n") #textovy retezec
    f.writelines(f"{i:>7},{i**2:>7},{i**0.5:>8.3f}\n" for i in range(1,11))

# Domácí úkol:
Pomocí vlastních tříd a dědičnosti vytvořte "soubor tříd" = nástroj pro derivaci funkcí.

Konkrétněji:
- Vytvořte třídu **Operator**
    - Třída bude mít metody:
        - `__init__()` - inicializuje operátor (zde stačí `pass`)
        - `__str__()` - vrací řetězec reprezentující operátor
        - `__repr__()` - vrací řetězec reprezentující operátor
- Vytvořte třídu **BinaryOperator**, která bude dědit od třídy Operator
    - Třída bude mít atributy:
        - `levy_argument` - levý operand
        - `pravy_argument` - pravý operand
    - Třída bude mít metody:
        - `__init__()` - inicializuje operátor, přiřadí argumenty
        - `__str__()` - vrací řetězec reprezentující operátor
        - `__repr__()` - vrací řetězec reprezentující operátor
        - `diff()` - vrací derivaci operátoru
- Vytvořte třídu **UnaryOperator**, která bude dědit od třídy Operator
    - Třída bude mít atributy:
        - `argument` - argument funkce 
    - Třída bude mít metody:
        - `__init__()` - inicializuje operátor, přiřadí argument
        - `__str__()` - vrací řetězec reprezentující operátor
        - `__repr__()` - vrací řetězec reprezentující operátor
        - `diff()` - vrací derivaci operátoru

Pro počítání derivací použijte standardní pravidla:
- derivace součtu je součet derivací
- pravidlo pro derivaci součinu
- derivace funkce je derivovaná funkce násobená derivovaným argumentem

Pro **BinaryOperator** a **UnaryOperator** implementujte také metody implementující standardní operátory:
- `__add__()` - implementuje operátor `+`
- `__sub__()` - implementuje operátor `-`
- `__mul__()` - implementuje operátor `*`

Pro **BinaryOperator** implementujte dvě dceřiné třídy:
- **Addition** - implementuje operátor `+`
- **Multiplication** - implementuje operátor `*`

Pro **UnaryOperator** implementujte tyto dceřiné třídy:
- **exp** - implementuje funkci `exp`
- **sin** - implementuje funkci `sin`
- **cos** - implementuje funkci `cos`
- **ln** - implementuje funkci `ln`
- **frc** - implementuje funkci `1/x`
- **identity** - implementuje identitu `x`
- **cst** - implementuje konstantu `c`
- **mocnina** - implementuje mocninu `x^n`

U všech funckí implementujte adekvátně metodu `diff()` a `__str__()`, tak aby se funkce vypisovaly v přehledné podobě.

Ozkoušejte na následujících funkcích:
- $sin(x + cos(x*x)) + 1$
- $x^2 + 2x + 1 + 2x^2$
- $x^2 + 2x + exp(x)$

In [17]:
# kostra k doplnění, místa označená TODO doplňte

class Operator:
    def __init__(self):
        pass
    
    def __str__(self):
        return type(self).__name__
    
    def __repr__(self):
        return self.__str__()
    
    def diff(self):
        pass

    def simplify(self):
        return self
    
    def __add__(self, other):
        return Addition(self, other)
    
    def __sub__(self, other):
        return Addition(self, Multiplication(cst(-1), other))
    
    def __mul__(self, other):
        return Multiplication(self, other)
    
class BinaryOperator(Operator):
    def __init__(self, levy_argument, pravy_argument):
        self.levy_argument = levy_argument
        self.pravy_argument = pravy_argument
        
class UnaryOperator(Operator):
    def __init__(self, argument):
        self.argument = argument
    
    #sin and stuff couldn't write what's inside them, added these
    def __str__(self):
        return f"{type(self).__name__}({self.argument})"
    def __repr__(self):
        return self.__str__()
    
        
class Addition(BinaryOperator):
    def __str__(self):
        return f"{self.levy_argument} + {self.pravy_argument}"
    def diff(self):
        return self.levy_argument.diff() + self.pravy_argument.diff()
    def simplify(self):
        left = self.levy_argument.simplify()
        right = self.pravy_argument.simplify()
        if type(left) == cst and type(right) == cst:
            left.value += right.value
            right.value = 0
        if type(left) == cst and left.value == 0:
            return right
        if type(right) == cst and right.value == 0:
            return left
        if left == right:
            return cst(2) * left
        return left + right
    
class Multiplication(BinaryOperator):
    def __str__(self):
        return f"({self.levy_argument})*({self.pravy_argument})"
    def diff(self):
        return self.levy_argument.diff() * self.pravy_argument + self.levy_argument * self.pravy_argument.diff()
    def simplify(self):
        left = self.levy_argument.simplify()
        right = self.pravy_argument.simplify()
        if type(left) == cst and type(right) == cst:
            left.value *= right.value
            right.value = 1
        if type(left) == cst:
            if left.value == 0:
                return cst(0)
            elif left.value == 1:
                return right
        if type(right) == cst:
            if right.value == 0:
                return cst(0)
            elif right.value == 1:
                return left
        if left == right:
            return mocnina(left, 2)
        return left * right
    
class cst(UnaryOperator):
    def __init__(self, value):
        self.value = value
    def __str__(self):
        return str(self.value)
    def diff(self):
        return cst(0)
    
class identity(UnaryOperator):
    def __init__(self):
        pass
    def __str__(self):
        return "x"
    def diff(self):
        return cst(1)
    
class mocnina(UnaryOperator):
    def __init__(self, argument, exponent):
        self.argument = argument
        self.exponent = exponent
    def __str__(self):
        return f"({self.argument})^{self.exponent}"
    def diff(self):
        if self.exponent == 0:
            return cst(0)
        elif self.exponent == 1:
            return self.argument
        return cst(self.exponent) * mocnina(self.argument, self.exponent-1) * self.argument.diff()
    def simplify(self):
        arg = self.argument.simplify()
        if self.exponent == 0:
            return cst(1)
        elif self.exponent == 1:
            return arg
        return mocnina(arg, self.exponent)
    
class sin(UnaryOperator):
    def diff(self):
        return cos(self.argument) * self.argument.diff()

class cos(UnaryOperator):
    def diff(self):
        return cst(-1) * sin(self.argument) * self.argument.diff()

class exp(UnaryOperator):
    def diff(self):
        return self.argument.diff() * exp(self.argument)
    
class ln(UnaryOperator):
    def diff(self):
        return self.argument.diff() * frc(self.argument)

class frc(UnaryOperator):
    def __str__(self):
        return f"1/({self.argument})"
    def diff(self):
        return self.argument.diff() * frc(mocnina(self.argument, 2))

In [18]:
# takto by se to mělo používat
x = identity()
# sin(x + cos(x*x)) + 1
f1 = sin(x + cos(x*x)) + cst(1)
print(f1)
print(f1.diff())
print((f1.diff()).simplify())

sin(x + cos((x)*(x))) + 1
(cos(x + cos((x)*(x))))*(1 + ((-1)*(sin((x)*(x))))*((1)*(x) + (x)*(1))) + 0
(cos(x + cos((x)*(x))))*(1 + ((-1)*(sin((x)*(x))))*((2)*(x)))


můj výstup byl:

`sin(x + cos((x)*(x))) + 1`

`(cos(x + cos((x)*(x))))*(1 + ((-1)*(sin((x)*(x))))*((1)*(x) + (x)*(1))) + 0`

In [19]:
x = identity()
# x ^ 2 + 2x + 1 + 2x ^ 2
f2 = mocnina(x,2) + cst(2)*x + cst(1) + cst(2)*mocnina(x,2)
print(f2)
print(f2.diff())
print((f2.diff()).simplify())

(x)^2 + (2)*(x) + 1 + (2)*((x)^2)
((2)*((x)^1))*(1) + (0)*(x) + (2)*(1) + 0 + (0)*((x)^2) + (2)*(((2)*((x)^1))*(1))
(2)*(x) + 2 + (2)*((2)*(x))


můj výstup byl:

`(x)^2 + (2)*(x) + 1 + (2)*((x)^2)`

`((2)*((x)^1))*(1) + (0)*(x) + (2)*(1) + 0 + (0)*((x)^2) + (2)*(((2)*((x)^1))*(1))`

In [20]:
x = identity()
# x ^ 2 + 2x + exp(x)
f3 = mocnina(x,2) + cst(2)*x + exp(x)
print(f3)
print(f3.diff())
print(f3.diff().simplify())

(x)^2 + (2)*(x) + exp(x)
((2)*((x)^1))*(1) + (0)*(x) + (2)*(1) + (1)*(exp(x))
(2)*(x) + 2 + exp(x)


můj výstup byl:

`(x)^2 + (2)*(x) + exp(x)`

`((2)*((x)^1))*(1) + (0)*(x) + (2)*(1) + (exp(x))*(1)`


# Bonusový úkol:
Doplňte ke všem třídám metodu `simplify()`, která provede zjednodušení výrazu. Například:
- součet s nulou vrátí druhý operand
- součin s nulou vrátí nulu
- mocnina s exponentem 0 vrátí 1
- mocnina s exponentem 1 vrátí první argument
- součín dvou stejných argumentů vrátí mocninu s exponentem 2
- součin konstant vrátí konstantu se součinem
- ...